### PR TITLE
Add metrics-deliver for tomte

### PR DESCRIPTION
**Pull request by fevac**
### Added

- metrics-deliver subcommand for tomte

### How to prepare for test

- [ ] Ssh to relevant server (depending on type of change)
- [ ] Use stage: `us`
- [ ] Paxa the environment: `paxa`
- [ ] Install on stage (example for Hasta):
    ```shell
    bash /home/proj/production/servers/resources/hasta.scilifelab.se/update-tool-stage.sh -e S_cg -t cg -b tomte_metrics -a
    ```

### How to test

- [x] dry-run `cg workflow tomte metrics-deliver -d CASE`
image

- [x] run poor QC case `cg workflow tomte metrics-deliver CASE`. Make sure that QC fails and command is aborted and that is reported in cigrid/trailblazer with a comment including the metrics that failed and status FAILED

image

image

- [x] run high QC case. As we currently do not have a case passing QC for tomte, first modify the multiqc_data.json to make all metrics pass and simulate a case passing QC. Then run `cg workflow tomte metrics-deliver CASE`. Make sure that QC pass and command completes successfully

image

image

## Review

- [x] Tests executed by EFC
- [ ] "Merge and deploy" approved by
Thanks for filling in who performed the code review and the test!

### This [version](https://semver.org/) is a

- [ ] **MAJOR** - when you make incompatible API changes
- [ ] **MINOR** - when you add functionality in a backwards compatible manner
- [ ] **PATCH** - when you make backwards compatible bug fixes or documentation/instructions

## Implementation Plan

- [ ] Document in ...
- [ ] Deploy this branch on ...
- [ ] Inform to ...

### CODE REVIEW

File: cg/meta/workflow/rnafusion.py [hunk 1]
```diff
@@ -111,30 +107,15 @@ def get_references_path(self, genomes_base: Path | None = None) -> Path:
...
+        """Get search patterns for multiqc for Rnafusion."""
+        return {
+            case_id: sample.internal_id
+            for sample in self.status_db.get_samples_by_case_id(case_id=case_id)
+        }
 
     @staticmethod
-    def ensure_mandatory_metrics_present(metrics: list[MetricsBase]) -> None:
+    def ensure_mandatory_metrics_present(metrics: set[MetricsBase]) -> None:
         """Check that all mandatory metrics are present. Raise error if missing."""
```
Comment (henrikstranneheim) [inline cg/meta/workflow/rnafusion.py:119]:
```suggestion
        """Check that all mandatory metrics are present. 
        Raise:  MissingMetrics if mandatory metrics are missing."""
```

File: cg/meta/workflow/taxprofiler.py [hunk 2]
```diff
@@ -96,35 +99,12 @@ def get_workflow_parameters(self, case_id: str) -> TaxprofilerParameters:
...
-        metrics_values: dict = {}
-        for stat_dict in multiqc_json:
-            for sample_key, sample_values in stat_dict.items():
-                if sample_key == f"{sample_name}_{sample_name}":
-                    LOG.info(f"Key: {sample_key}, Values: {sample_values}")
-                    metrics_values.update(sample_values)
-
-        return metrics_values
+    def multiqc_search_patterns(self, case_id: str) -> dict:
+        """Get search patterns for multiqc for Taxprofiler."""
```
Comment (henrikstranneheim) [inline cg/meta/workflow/taxprofiler.py:103]:
```suggestion
    def get_multiqc_search_pattern(self, case_id: str) -> dict:
        """Return search patterns for MultiQC."""
```

File: tests/cli/workflow/nf_analysis/test_cli_metrics_deliver.py [hunk 3]
```diff
@@ -36,7 +38,7 @@ def test_metrics_deliver_without_options(
 
 @pytest.mark.parametrize(
     "workflow",
-    [Workflow.RNAFUSION, Workflow.TAXPROFILER],
+    [Workflow.RNAFUSION, Workflow.TAXPROFILER, Workflow.TOMTE],
```
Comment (henrikstranneheim) [inline tests/cli/workflow/nf_analysis/test_cli_metrics_deliver.py:41]: So nice and easy, 💯

**Review by henrikstranneheim — APPROVED** [review]
Beautiful, 👍

File: cg/meta/workflow/nf_analysis.py [hunk 4]
```diff
@@ -587,26 +598,80 @@ def get_workflow_metrics(self) -> dict:
...
+
+    def parse_multiqc_json_for_pattern(
+        self,
+        search_pattern: str,
+        multiqc_json: MultiqcDataJson,
+        metric_id: str,
+        exact_match: bool = False,
+    ) -> list[MetricsBase]:
+        """Parse a MultiqcDataJson and returns a list of metrics."""
+        metrics_values = []
```
Comment (ivadym) [inline cg/meta/workflow/nf_analysis.py:647]:
```suggestion
        metrics: list[MetricsBase] = []
```

**Review by ivadym — COMMENTED** [review]
Great job! Love how you're integrating all the workflows and reusing the common code 💯

File: cg/meta/workflow/nf_analysis.py [hunk 5]
```diff
@@ -587,26 +598,80 @@ def get_workflow_metrics(self) -> dict:
...
 
+    def multiqc_search_patterns(self, case_id: str) -> dict:
+        """Get search patterns for multiqc. Each key is a search pattern and each value
+        corresponds to the metric id to set in the metrics deliverables file.
+        Multiple search patterns can be added, and ideally it should be
+        {sample_id_1: sample_id_1, sample_id_2: sample_id_2}."""
+        return {
+            sample_id: sample_id
+            for sample_id in self.status_db.get_sample_ids_by_case_id(case_id=case_id)
+        }
```
[the remark this answers is not in the document]
Comment (fevac) [reply cg/meta/workflow/nf_analysis.py:609]: IMO it's quite readable but I changed it now!

File: cg/meta/workflow/nf_analysis.py [hunk 6]
```diff
@@ -587,26 +598,80 @@ def get_workflow_metrics(self) -> dict:
         """Get nf-core workflow metrics constants."""
         return {}
 
+    def multiqc_search_patterns(self, case_id: str) -> dict:
+        """Get search patterns for multiqc. Each key is a search pattern and each value
+        corresponds to the metric id to set in the metrics deliverables file.
+        Multiple search patterns can be added, and ideally it should be
+        {sample_id_1: sample_id_1, sample_id_2: sample_id_2}."""
```
[the remark this answers is not in the document]
Comment (fevac) [reply cg/meta/workflow/nf_analysis.py:605]: `ideally it should be` --> ` ideally in the format`:  it not about the format, it's about which patterns we use. I will skip that change but I rephrased it to make it clearer

File: cg/meta/workflow/nf_analysis.py [hunk 7]
```diff
@@ -587,26 +598,80 @@ def get_workflow_metrics(self) -> dict:
...
+        return {
+            sample_id: sample_id
+            for sample_id in self.status_db.get_sample_ids_by_case_id(case_id=case_id)
+        }
+
+    def remove_duplicated_metrics(self, metrics: list[MetricsBase]) -> list[MetricsBase]:
```
[the remark this answers is not in the document]
Comment (fevac) [reply cg/meta/workflow/nf_analysis.py:611]: already updated based on Henrik suggestion to `get_deduplicated_metrics`

File: cg/meta/workflow/nf_analysis.py [hunk 8]
```diff
@@ -587,26 +598,80 @@ def get_workflow_metrics(self) -> dict:
...
+        """Remove duplicated metrics based on id and sample name. If duplicated entries are found
```
[the remark this answers is not in the document]
Comment (fevac) [reply cg/meta/workflow/nf_analysis.py:612]: Updated. Not sure what's the consensus if ID or id, I'm trying to get an answer and then will modify if needed

File: cg/meta/workflow/nf_analysis.py [hunk 9]
```diff
@@ -587,26 +598,80 @@ def get_workflow_metrics(self) -> dict:
...
+        for search_pattern, metric_id in self.multiqc_search_patterns(case_id=case_id).items():
+            metrics_for_pattern: list[MetricsBase] = self.parse_multiqc_json_for_pattern(
+                search_pattern=search_pattern,
+                multiqc_json=multiqc_json,
+                metric_id=metric_id,
+                exact_match=self.is_multiqc_pattern_search_exact,
             )
-        return metric_base_list
+            metrics.extend(metrics_for_pattern)
+        metrics = self.remove_duplicated_metrics(metrics=metrics)
```
[the remark this answers is not in the document]
Comment (fevac) [reply cg/meta/workflow/nf_analysis.py:636]: the most common case I see is metrics that are calculated separately for read 1  (forward) and read 2 (reverse). Then we have the mean length, gc content, etc for both. In the original multiqc file they usually appear with a suffix (*_1 and *_2) but currently we are not parsing the suffixes and only getting one of the metrics. They usually have very very similar values for _1 and _2 but we could change this behaviour in the future and parse both, or get an average, or something else

File: cg/meta/workflow/taxprofiler.py [hunk 10]
```diff
@@ -60,6 +58,11 @@ def is_multiple_samples_allowed(self) -> bool:
         """Return whether the analysis supports multiple samples to be linked to the case."""
         return True
 
+    @property
+    def is_multiqc_pattern_search_exact(self) -> bool:
```
[the remark this answers is not in the document]
Comment (fevac) [reply cg/meta/workflow/taxprofiler.py:62]: what is exact is the search not the pattern. leaving as it is

File: cg/meta/workflow/nf_analysis.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (fevac) [reply cg/meta/workflow/nf_analysis.py:612]: updated to ID

File: cg/meta/workflow/nf_analysis.py [hunk 11]
```diff
@@ -587,26 +598,80 @@ def get_workflow_metrics(self) -> dict:
...
-                    header=None,
-                    id=sample_id,
-                    input=MultiQC.MULTIQC_DATA + FileExtensions.JSON,
-                    name=metric_name,
-                    step=MultiQC.MULTIQC,
-                    value=metric_value,
-                    condition=self.get_workflow_metrics().get(metric_name, None),
-                )
+        multiqc_json: MultiqcDataJson = MultiqcDataJson(
+            **read_json(file_path=self.get_multiqc_json_path(case_id=case_id))
```
[the remark this answers is not in the document]
Comment (fevac) [reply cg/meta/workflow/nf_analysis.py:625]: model_validate is not supported by pydantic.v1 - so only excluding type hints here
pydantic version for nf workflows should be updated in a different PR

**Comment by fevac** [comment]
tests are passing

File: cg/meta/workflow/nf_analysis.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (fevac) [reply cg/meta/workflow/nf_analysis.py]: resolving

**Comment by fevac** [comment]
- deployed

image

image

**Pull request merged**